web/packages/teleport/src/Bots/Edit/EditDialog.test.tsx

```diff
@@ -0,0 +1,446 @@
+    fireEvent.click(saveButton);
+    await waitForSaveButton();
+
+    expect(onSuccess).toHaveBeenCalledTimes(1);
+    expect(onSuccess).toHaveBeenLastCalledWith(
+      {
+        description: '',
+        kind: 'bot',
+        labels: new Map(),
+        max_session_ttl: {
+          seconds: 43200 + 30 * 60,
+        },
+        name: 'test-bot',
+        namespace: '',
+        revision: '',
+        roles: ['admin', 'user'],
+        status: 'active',
+        subKind: '',
+        traits: [
+          {
+            name: 'trait-1',
+            values: ['value-1', 'value-2', 'value-3'],
+          },
+        ],
+        type: null,
+        version: 'v1',
+      },
+      false
+    );
+  });
+
+  it('should show a save error state', async () => {
+    const onSuccess = jest.fn();
+
+    withFetchBotSuccess();
+    withFetchRolesSuccess();
+    renderComponent({ onSuccess });
+    await waitForLoading();
+
+    // Change something to enable the save button
+    await inputMaxSessionDuration('12h 30m');
+
+    withSaveError();
+    const saveButton = screen.getByRole('button', { name: 'Save' });
+    expect(saveButton).toBeEnabled();
+    fireEvent.click(saveButton);
+    await waitForSaveButton();
+
+    expect(screen.getByText('Error: something went wrong')).toBeInTheDocument();
+
+    expect(onSuccess).not.toHaveBeenCalled();
+  });
+
+  it('should show a save inconsistency warning', async () => {
+    const onSuccess = jest.fn();
+
+    withFetchBotSuccess();
+    withFetchRolesSuccess({ items: ['test-role'] });
+    renderComponent({ onSuccess });
+    await waitForLoading();
+
+    await inputRole('test-role');
+    await inputTrait('logins', ['test-value']);
+    await inputMaxSessionDuration('12h 30m');
+
+    // Mock the server response to deliberately not match the request, triggering the warning on all fields
+    withSaveSuccess({
+      roles: ['admin', 'user'],
+      traits: [
+        {
+          name: 'trait-1',
+          values: ['value-1', 'value-2', 'value-3'],
+        },
+      ],
+      max_session_ttl: '12h',
+    });
+    const saveButton = screen.getByRole('button', { name: 'Save' });
+    expect(saveButton).toBeEnabled();
+    fireEvent.click(saveButton);
+    await waitForSaveButton();
+
+    expect(onSuccess).toHaveBeenCalled();
+
+    expect(
+      screen.getByText(
+        'Warning: Some fields may not have updated correctly; max_session_ttl, roles, traits'
+      )
+    ).toBeInTheDocument();
+  });
+});
+
+async function inputRole(role: string) {
+  const rolesSelect = screen.getByLabelText('Roles');
+  // Seems to be the only way to interact with react-select component
+  fireEvent.keyDown(rolesSelect, { key: 'ArrowDown' });
+  const item = await screen.findByText(role);
+  fireEvent.click(item);
+}
+
+async function inputTrait(name: string, values: string[]) {
+  const traitSelect = screen.getAllByLabelText('Key');
+  // Seems to be the only way to interact with react-select component
```

Comment:
I know that sometimes we use `react-select-event` package, maybe it could help here?